py/htm/encoders/eye.py

```diff
@@ -0,0 +1,479 @@
+        Use the result to determine the actual label of the image in the area
+        where the eye is looking.
+        """
+        # Find the retina's radius in the image.
+        r = int(round(self.scale * self.retina_diameter / 2))
+        # Shrink the retina's radius so that sample points are nearer the fovea.
+        # Also shrink radius B/C this does not account for the diagonal
+        # distance, just the manhattan distance.
+        r = r * 2/3
+        # Generate points.
+        coords = np.random.random_integers(-r, r, size=(npoints, 2))
+        # Add this position offset.
+        coords += np.array(np.rint(self.position), dtype=np.int).reshape(1, 2)
+        return coords
+
+    def small_random_movement(self):
+        """returns small difference in position, rotation, scale.
+           This is naive "saccadic" movements.
+        """
+        max_change_angle = (2*3.14159) / 500
+        self.position = (
+            self.position[0] + random.gauss(1, .75),
+            self.position[1] + random.gauss(1, .75),)
+        self.orientation += random.uniform(-max_change_angle, max_change_angle)
+        self.scale = 1
+        return (self.position, self.orientation, self.scale)
+
+
+    def reset(self):
+        self.retina.clearBuffers()
+
+
+
+def _get_images(path):
+    """ Returns list of all image files found under the given file path. """
+    image_extensions = [
+        '.bmp',
+        '.dib',
+        '.png',
+        '.jpg',
+        '.jpeg',
+        '.jpe',
+        '.tif',
+        '.tiff',
+    ]
+    images = []
+    import os
+    if os.path.isfile(path):
+        basename, ext = os.path.splitext(path)
+        if ext.lower() in image_extensions:
+            images.append( path )
+    else:
+        for dirpath, dirnames, filenames in os.walk(path):
+            for fn in filenames:
+                basename, ext = os.path.splitext(fn)
+                if ext.lower() in image_extensions:
+                    images.append( os.path.join(dirpath, fn) )
+    return images
+
+if __name__ == '__main__':
+    import argparse
+    args = argparse.ArgumentParser()
+    args.add_argument('IMAGE', type=str)
+    args   = args.parse_args()
+    images = _get_images( args.IMAGE )
+    random.shuffle(images)
+    if not images:
+        print('No images found at file path "%s"!'%args.IMAGE)
+    else:
+        eye = Eye()
+        for img_path in images:
+            eye.reset()
+            print("Loading image %s"%img_path)
+            eye.new_image(img_path)
+            eye.scale = 1
+            for i in range(10):
+                pos,rot,sc = eye.small_random_movement()
+                sdr = eye.compute(pos,rot,sc) #TODO derive from Encoder
+                eye.show_view()
+                print("Sparsity: {}".format(len(sdr.sparse)/np.product(sdr.dimensions)))
+                print(sdr.dimensions) #TODO make SDR 2D, diameter x diameter. not current (200, 200, 2) 
```

Comment:
TODO make SDR 2D, diameter x diameter. not current (200, 200, 2). 

the 3rd dimensions represents P,M-cells? In that case, we should either: 
- drop the `mode` argument
- or keep mode and merge the SDR into a single layer (2D) 
  - I don't know if biologically the P,M-pathways are merged together, or lead to different regions?

Reply:
> But does not conform.

That sounds like a bug!

---

Yes, the 3rd dimension is for P&M cells.

> I don't know if biologically the P,M-pathways are merged together, or lead to different regions?

IIRC: They project to different things in the midbrain & thalamus.  I'm not sure about what happens in the cortex.

Reply:
> > But does not conform.

> That sounds like a bug!

Resolved, my bad, sorry! I computed the sparsity from both (the 3rd dimension) SDRs, now the computation is fixed and it gives 20% as expected. 

> Yes, the 3rd dimension is for P&M cells.

For clarity I added members 
```
self.parvo_sdr
self.magno_sdr
self.parvo_img
self.magno_img
```